### PR TITLE
vault12-integration-pt2

### PR DESCRIPTION
**Pull request by lomchik**

### CODE REVIEW

File: src/crypto-storage/crypto-storage.ts [hunk 1]
```diff
@@ -55,10 +55,9 @@ export class CryptoStorage {
 
   static setStorageDriver(driver: StorageDriver) {
     if (this.storageDriver) {
-      throw new Error('[NaCl] NaCl driver has been already set, it is supposed to be set only once');
-    } else {
-      this.storageDriver = driver;
+      console.warn('[CryptoStorage] StorageDriver has been already set, it is supposed to be set only once');
```
Comment (pavlo-liapin) [inline src/crypto-storage/crypto-storage.ts:58]: Why do you want to let users overwrite the driver? I thought it should be set only once, to avoid stupid mistakes

File: src/nacl/nacl.ts [hunk 2]
```diff
@@ -16,11 +16,9 @@ export class NaCl {
 
   public static setInstance(driver: NaClDriver): boolean {
     if (this.driverInstance) {
-      throw new Error('[NaCl] NaCl driver has been already set, it is supposed to be set only once');
-    } else {
-      // fallback to the default JS driver
-      this.driverInstance = driver;
+      console.warn('[NaCl] NaCl driver has been already set, it is supposed to be set only once');
```
Comment (pavlo-liapin) [inline src/nacl/nacl.ts:19]: Same question, why let users overwrite?

File: src/zax.interface.ts [hunk 3]
```diff
@@ -80,6 +80,7 @@ export interface ZaxPlainMessage {
   time: number;
   from: Base64;
   nonce: Base64;
+  senderTag: undefined;
```
Comment (pavlo-liapin) [inline src/zax.interface.ts:83]: Why is this needed? Is it for using `senderTag` field of `ZaxParsedMessage` type on the client?

File: src/mailbox/mailbox.offline.spec.ts [hunk 4]
```diff
@@ -66,7 +66,7 @@ describe('Mailbox / Offline tests', () => {
   });
 
   it('encrypts raw binary data', async () => {
```
Comment (pavlo-liapin) [inline src/mailbox/mailbox.offline.spec.ts:68]: Please rename this test, it doesn't match contents anymore. Maybe `it('produces proper encrypted messages after encryption')`. Or just move `.toHaveLength()` checks to the previous test and remove this one

File: src/mailbox/mailbox.transfer-messages.spec.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+describe('Mailbox / Transfer Messages', () => {
+
+  let Alice: Mailbox;
+  let Bob: Mailbox;
+  const messages = [
+    JSON.stringify({object: 'message'}),
+    'some unencrypted message',
+    'special ;@@#2sd characters',
+    'кирилиця'
+  ].slice(0,4);
```
Comment (pavlo-liapin) [inline src/mailbox/mailbox.transfer-messages.spec.ts:16]: Just `.slice()` is fine too, output is the same

**Review by pavlo-liapin — APPROVED** [review]

File: src/crypto-storage/crypto-storage.ts (hunk 1 above)
Comment (lomchik) [reply src/crypto-storage/crypto-storage.ts:58]: for testing purpose some times we need to reset driver

File: src/zax.interface.ts (hunk 3 above)
Comment (lomchik) [reply src/zax.interface.ts:83]: yep

File: src/mailbox/mailbox.ts [hunk 6]
```diff
@@ -349,21 +354,30 @@ export class Mailbox {
...
    */
-  async decodeMessage(guest: string, nonce: Base64, ctext: Base64): Promise<any> {
+  async decodeMessage(guest: string, nonce: Base64, ctext: Base64) {
     const guestPk = this.getGuestKey(guest);
     const privateKey = this.keyRing.getPrivateCommKey();
+    let uint8ArrayCtext: Uint8Array;
+    try {
+      uint8ArrayCtext = Utils.fromBase64(ctext);
+    } catch (err) {
+      // looks like ctext was not encoded
```
Comment (dmitry-salnikov) [inline src/mailbox/mailbox.ts:376]: does it make sense to throw error with this explanation or at least `console.log` it?

File: src/nacl/encryption.helper.ts [hunk 7]
```diff
@@ -13,31 +13,26 @@ export class EncryptionHelper {
   /**
    * Encodes a binary message with `cryptobox`
    */
-  static async encodeMessage(message: any, pkTo: Uint8Array, skFrom: Uint8Array, nonceData?: number):
-    Promise<EncryptedMessage> {
+  static async encodeMessage(message: Uint8Array, pkTo: Uint8Array, skFrom: Uint8Array, nonceData?: number) {
```
Comment (dmitry-salnikov) [inline src/nacl/encryption.helper.ts:16]: why did you remove return types here and below/above?

**Review by dmitry-salnikov — APPROVED** [review]

File: src/nacl/nacl.ts (hunk 2 above)
Comment (lomchik) [reply src/nacl/nacl.ts:19]: this is also for testing purpose

File: src/mailbox/mailbox.offline.spec.ts (hunk 4 above)
Comment (lomchik) [reply src/mailbox/mailbox.offline.spec.ts:68]: renamed

File: src/mailbox/mailbox.transfer-messages.spec.ts (hunk 5 above)
Comment (lomchik) [reply src/mailbox/mailbox.transfer-messages.spec.ts:16]: removed I was using it for testing

File: src/mailbox/mailbox.ts [hunk 8]
```diff
@@ -349,21 +354,30 @@ export class Mailbox {
    * Encodes a free-form object `message` to the guest key of a guest already
    * added to the keyring
    */
-  async encodeMessage(guest: string, message: any): Promise<EncryptedMessage> {
+  async encodeMessage(guest: string, message: string): Promise<EncryptedMessage> {
```
[the remark this answers is not in the document]
Comment (lomchik) [reply src/mailbox/mailbox.ts:357]: done

File: src/mailbox/mailbox.ts (hunk 6 above)
Comment (lomchik) [reply src/mailbox/mailbox.ts:376]: nope no reason

File: src/nacl/encryption.helper.ts (hunk 7 above)
Comment (lomchik) [reply src/nacl/encryption.helper.ts:16]: caus e it defined by value it return

File: src/crypto-storage/crypto-storage.ts (hunk 1 above)
Comment (lomchik) [reply src/crypto-storage/crypto-storage.ts]: done

File: src/nacl/nacl.ts (hunk 2 above)
Comment (lomchik) [reply src/nacl/nacl.ts]: bring error throwing back

**Review by pavlo-liapin — APPROVED** [review]

**Review by dmitry-salnikov — APPROVED** [review]
